snuba/replacers/replacer_processor.py

```diff
@@ -16,14 +17,36 @@ class ReplacementMessage(NamedTuple):
     data: Mapping[str, Any]
 
 
-@dataclass(frozen=True)
-class Replacement:
-    # XXX: For the group_exclude message we need to be able to run a
-    # replacement without running any query.
-    count_query_template: Optional[str]
-    insert_query_template: Optional[str]
-    query_args: Mapping[str, Any]
-    query_time_flags: Any
+class Replacement(ABC, Generic[C]):
+    @abstractclassmethod
+    def parse_message(
+        cls, message: Mapping[str, Any], context: C
+    ) -> Optional["Replacement[C]"]:
```

Comment:
I am not sure I get why you need to make this class generic to the type of context.  It seems you do not care about that type in any of the interface method except for the parser. So it seems the type of context the class operates on could be an implementation detail of the concrete classes with no need for this to be generic. As a consequence you would not have to reference the specific parameter type every time you reference `Replacement`.

Also you do not need to specify the `parse_message` method in the abstract class. The code that instantiates the Replacement object knows the concrete classes. It can reference the specific parse_message method.

Reply:
You are right, if I remove `parse_message` from the ABC I have removed all exposure of context to `replacer_processor`.

I may have to write an additional ABC for errors_replacer, put parse_message there and also get_excluded_groups as both are errors_replacer specific